clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -308,11 +308,12 @@ static void reportConflictingAttrs(Sema &S, FunctionDecl *F, const Attr *A1,
   F->setInvalidDecl();
 }
 
-// Returns the signed constant integer value represented by given expression.
-static int64_t getIntExprValue(Sema &S, const Expr *E) {
+/// Returns the signed constant integer value represetned by given expression
+static int getIntExprValue(const Expr *E, ASTContext &Ctx) {
   llvm::APSInt Val(32);
-  bool IsValid = E->isIntegerConstantExpr(Val, S.getASTContext());
+  bool IsValid = E->isIntegerConstantExpr(Val, Ctx);
   assert(IsValid && "expression must be constant integer");
+  (void)IsValid;
```

Comment:
This stuff doesn't look good.
```suggestion
  assert(E->isIntegerConstantExpr(Val, Ctx) && "expression must be constant integer");
```

Reply:
This is not correct: `isIntegerConstantExpr` performs evaluation and stores result to `Val` (a bit unexpected, judging by the name). Having this within `assert` statement will remove evaluation from builds without assertions (default Release build)

Reply:
Yep, exactly.  We need the IsValid/assert/void-cast because the 'assert' is actually a macro.

This means:
1- The "isValid" line needs to work in both asserts and no asserts mode, so it has to be its own line.
2- We want the assert for asserts builds.
3- To suppress an 'unused variable' warning in non-asserts mode, we do the void-cast.

Reply:
I was under impression that `isIntegerConstantExpr` doesn't perform any evaluations, so `isValid` is not needed as a separate line. Thanks for the explanation, resolved.